### PR TITLE
*: delete a bunch of generated go code from tree 

### LINKED ISSUE #72232: dev: spurious diffs appear when hoisting generated code into tree
**Describe the problem**

The checked in generated code (protos and various stringer files) are generated using `make`. When generating using bazel/dev (introduced in https://github.com/cockroachdb/cockroach/pull/68663), you can observe spurious diffs.
```
$ dev build cockroach-short --hoist-generated-code
$ git diff
```

The diffs (at the time of writing) are of three sorts:
```diff
diff --git i/pkg/base/testclusterreplicationmode_string.go w/pkg/base/testclusterreplicationmode_string.go
index 6380f66b3e..c14b8c058f 100644
--- i/pkg/base/testclusterreplicationmode_string.go
+++ w/pkg/base/testclusterreplicationmode_string.go
@@ -1,4 +1,4 @@
-// Code generated by "stringer -type=TestClusterReplicationMode"; DO NOT EDIT.
+// Code generated by "stringer -output=bazel-out/darwin-fastbuild/bin/pkg/base/testclusterreplicationmode_string.go -type=TestClusterReplicationMode pkg/base/test_server_args.go"; DO NOT EDIT.
```

```diff
diff --git i/pkg/ccl/baseccl/encryption_options.pb.go w/pkg/ccl/baseccl/encryption_options.pb.go
index a7edf44210..63e766c050 100644
--- i/pkg/ccl/baseccl/encryption_options.pb.go
+++ w/pkg/ccl/baseccl/encryption_options.pb.go
@@ -611,3 +611,4 @@ var (
        ErrIntOverflowEncryptionOptions          = fmt.Errorf("proto: integer overflow")
        ErrUnexpectedEndOfGroupEncryptionOptions = fmt.Errorf("proto: unexpected end of group")
 )
+
```

And new (go ignored) files:
```sh
$ cat pkg/server/serverpb/init.pb.gw.go
// +build ignore

package ignore
```

**Expected behavior**

For there to be no extraneous diffs. For protos the extra empty line is something `gofmt` gets rid of, so feels to me we should get `dev`/`bazel` to apply it when hoisting out of the sandbox. For the stringer files, we could either ignore them in the source tree or use a custom `stringer` binary in `bazel` to strip out the sandbox path? Not sure about the .pb.gw.go files.

Epic CRDB-8036

### PR DESCRIPTION
**Pull request by rickystewart**
Before this change, `dev build --hoist-generated-code` would introduce a
lot of spurious, benign diffs in generated code. I've validated that
you can `dev build --hoist-generated code` without introducing any
diffs with this change applied.

The code we're deleting from tree here is:
* *ALL* `.pb.go` and `.pb.gw.go` files;
* and an ad-hoc selection of generated files that cannot be hoisted into
  the worktree by `dev` without introducing diffs (see #72232).

My first attempt at solving this problem was by deleting "all" generated
code from tree but I quickly found that this was intractable: the large
majority of these generated files don't have bespoke support in the
`Makefile`, so deleting them from tree and running `make build` just
causes the build to fail. Instead, the approach here is more targeted:
I delete only the generated files that have diffs when hoisted by `dev`,
and add support in the `Makefile` for those specific generated files.
The remaining checked-in generated files aren't hurting anyone, and can
be deleted from tree wholesale after `make` is dead.

NB: This change can require running `make generate` (or, in certain more
constrained scenarios, `make protobuf`) to make some builds/tests work.

Closes #72232.

Release note: None

### CODE REVIEW

**Comment by rickystewart** [comment]
Ah, CI is failing. Give me a bit to resolve...

**Comment by otan** [comment]
Chiming in, but historically I recall we checked in generated files because in cases where e.g. they get deleted between versions and they didn't get checked in, they remain in the tree and cause weird build issues all over. I wonder whether `--hoist-generated-code` needs to get smart about removing generated files as well.

**Comment by rickystewart** [comment]
Ah, I'm realizing that the `Makefile` doesn't have appropriate rules for the `stringer` files, so we can't delete them. 🙄 I don't know that updating the `Makefile` to fix this is a possibility, but I think we can easily patch `stringer` to quash the diffs. I'll do that tomorrow.

> I wonder whether --hoist-generated-code needs to get smart about removing generated files as well.

Not sure that it needs to "get smart", but `--hoist-generated-code` could do a `git clean` (appropriately constrained so that we only clean `.go` files, excepting `node_modules`, etc.) before hoisting. This would have the same effect. https://github.com/cockroachdb/cockroach/issues/72678

**Comment by otan** [comment]
> could do a git clean (appropriately constrained so that we only clean .go files, excepting node_modules, etc.) 

figuring out which files to clean is the "smart" bit, that's some complex AI/ML work right there :D

**Comment by irfansharif** [comment]
Another option is to introduce a git pre branch change hook (is that a thing?) to clean up some of these now ignored files. I'm not sure how fast that would be. FWIW, ignoring new files and having them appear in old branches has been a problem for a while, but I don't think that it means we should have them checked in perpetually.

**Comment by knz** [comment]
I recall there's at least 1 generated file in pkg/roachpb that has a circular dependency with its generator; i.e. if we delete the file we can't re-generate it any more.  
(Don't ask me how it ever was put there in the first place...)

Tobias and I would both welcome an improvement in that area! But it's not just a matter of makefile rules.

**Comment by rickystewart** [comment]
> I recall there's at least 1 generated file in pkg/roachpb that has a circular dependency with its generator; i.e. if we delete the file we can't re-generate it any more.

That's probably `pkg/roachpb/batch_generated.go`. This PR has [an exception](https://github.com/cockroachdb/cockroach/pull/72675/files#diff-bc37d034bad564583790a46f19d807abfe519c5671395fd494d8cce506c42947R50) for that file, and it's also mentioned in `build/bazelutil/checked_in_genfiles.txt` as one of the titular checked-in genfiles.

**Comment by knz** [comment]
thanks for checking

**Comment by rickystewart** [comment]
https://github.com/cockroachdb/cockroach/pull/72704 for the stringer patch. #72699 for cleaning old ignored files from `dev`. I'll rebase and make sure CI passes with the change.

File: .gitignore
Comment (irfansharif) [inline .gitignore:49]: We have .gitignores within specific packages, perhaps these could go there?
```
$  find . -type f | grep gitignore
[...]
./sql/parser/.gitignore
./sql/opt/optgen/lang/support/vscode/cockroachdb.optgen-1.0.0/.gitignore
./sql/opt/xform/.gitignore
./sql/opt/norm/.gitignore
./sql/opt/.gitignore
./sql/opt/memo/.gitignore
```

File: Makefile
Comment (irfansharif) [inline Makefile:1452]: Do you mind explaining why these specific files are special cased? Are there other go:generate ones that aren't? What about the stringer files? See top-level comment -- do we plan to .gitignore other kinds of generated code shortly?

**Review by irfansharif — APPROVED** [review]
I assume this will break a lot of workflows in the short term. 
- If you were to switch to an older branch, you'll have these empty pb.gw.go files around;
- If you were to switch to an older branch and then switch back, you'd have a lot of stale pb.go files and pb.go-only directories[1] around (breaking builds opaquely);
- For IDEs to work, you'd still need to generate these files manually and place in your source tree (either through `dev` or `make {generate,protobuf}`);
- If there are in-flight branches rebased off of commits prior to this one, you might see merge skews.

It would help enumerating some of these scenarios in the commit message+body and the suggested work arounds (making it less `dev` specific would also help). I assume we're planning to send out a PSA to eng@/#engineering as well?

We also mentioned an exception for `pkg/roachpb/batch_generated.go` [earlier](https://github.com/cockroachdb/cockroach/pull/72675#issuecomment-967229796), but this PR no longer has it. I see that we aren't deleting that file from tree, and I guess a few others (_string.go) as well. Do we plan to?

[1]: `pkg/kv/kvserver/loqrecovery/` for e.g. 

---

Given the large surface area here, we're sure to have merge skews (I see we have a conflict already). I'll go ahead and rubber stamp so you're able to land it whenever most (in)convenient.

**Comment by irfansharif** [comment]
A few extra files to consider:
```
$ dev build short --hoist-generated-code
WARNING: Ignoring JAVA_HOME, because it must point to a JDK, not a JRE.
INFO: Analyzed target //pkg/cmd/cockroach-short:cockroach-short (0 packages loaded, 0 targets configured).
INFO: Found 1 target...
Target //pkg/cmd/cockroach-short:cockroach-short up-to-date:
  _bazel/bin/pkg/cmd/cockroach-short/cockroach-short_/cockroach-short
INFO: Elapsed time: 2.582s, Critical Path: 0.40s
INFO: 1 process: 1 internal.
INFO: Build completed successfully, 1 total action
Successfully built binary for target //pkg/cmd/cockroach-short:cockroach-short at cockroach-short

$ git status
On branch gitignore_genfiles
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   pkg/cmd/roachtest/prometheus/mock_generated.go
        modified:   pkg/cmd/roachtest/tests/drt_generated.go
        modified:   pkg/roachprod/vm/aws/embedded.go
```

Do we want a CI check/test that ensures a clean slate?

**Comment by otan** [comment]
Sorry one more chime. 

This is probably going to be a somewhat painful crossover because of the generated files appearing and disappearing between branch swaps. Would probably be a good idea for an email to be sent out about how to deal with these files in common flows (Eg how to clear and build an older backport branch with all these automatic files in) and advising people to rebase all their branches. 

Sorry to add to the bureaucracy... want to make sure people have a positive experience going into bazel.

**Comment by tbg** [comment]
@irfansharif looks like you have a good handle on the UX issues that could come out of this branch, I would feel comfortable merging this when you've given this branch a whirl with your Goland setup & found it workable & we make sure all the new caveats are documented & we think they're acceptable.

At some point we should recommend everyone maintain multiple worktrees (one for each "generation" of build system or release branch) but right now you can't even `make build` unless the repo sits at $GOPATH/src/github.com/cockroachdb/cockroach. That will hopefully become better with bazel but first we have to get there.

**Comment by irfansharif** [comment]
Goland without running tests through it works just fine (my workflow), but I just tried running tests through it which I think @tbg and others frequent. Because Goland calls into `go test` directly, with this PR you'd would have to `make protobuf` to update pb.go files[1]. I think you'd have to do it every time you switch branches. Note that you wouldn't have to do it if you're running tests/builds through `make` or `bazel` because the proto dependencies are tracked properly and generated on the fly. I'm not sure what an work around there is, or if the caveat is acceptable. Is there a way to set up goland to run triggers before test runs? Or hook it up to `make test`? Should we generate protos + specific generated files  in tree as part of a post-branch-switch commit trigger?

[1]: Actually with the few extra .gitignored files in this PR (`pkg/security/securitytest/embedded.go`), you'd have to also generate them by hand/`make generate`/through Goland.

**Comment by tbg** [comment]
I think it's acceptable to `make protobuf` after switching branches before in-IDE `go test` works (btw how are you not using that? Do you never use the debugger?). It is already necessary today to invoke some `make` targets to regenerate the various generated files.

**Comment by tbg** [comment]
Goland also does allow you to edit the default "Go test" template and to attach pre-run triggers. If we have one that is cheap in steady state but takes care of the rebuild that would be something I'd want to document on the Goland wiki.

**Comment by irfansharif** [comment]
> btw how are you not using that? Do you never use the debugger?

I only use it for the debugger, but usually I just shell out to use dev/bazel. There's a Goland integration to run tests through bazel directly but I haven't tried it with the crdb repo. TODO...

**Comment by rickystewart** [comment]
@irfansharif Can you send me the `git diff` from that build run w/ `--hoist-generated-code`? Need to investigate those.

**Comment by rickystewart** [comment]
Oh, I assume they're from previous `dev` builds of `roachtest`/`roachprod`. I haven't tested those. Will look into it.

**Comment by irfansharif** [comment]
For posterity: [dev-gen-diff.txt](https://github.com/cockroachdb/cockroach/files/7763599/dev-gen-diff.txt).

File: .gitignore
Comment (rickystewart) [reply .gitignore:51]: Sorry, what's the request here? That I add new `.gitignore`s for the packages referenced here and add these files to those `.gitignore`s? If it's up to me, I really prefer having this stuff in one place, but I'll make that change if we're adamant about it.

**Comment by rickystewart** [comment]
I updated the commit message/PR description, should be a little easier to wrap your head around.

> If you were to switch to an older branch, you'll have these empty pb.gw.go files around;

If it makes a difference, I'd be happy to add these `.pb.gw.go` files to `.gitignore` for old branches.

> At some point we should recommend everyone maintain multiple worktrees (one for each "generation" of build system or release branch) but right now you can't even `make build` unless the repo sits at $GOPATH/src/github.com/cockroachdb/cockroach. That will hopefully become better with bazel but first we have to get there.

The Bazel build shouldn't have any similar restrictions, though I haven't tested it. Jury's out for any helper scripts or whatever that assumes the existence of the `GOPATH`.

File: .gitignore
Comment (irfansharif) [reply .gitignore:51]: Not at all adamant -- excited to see this go in!

**Comment by rickystewart** [comment]
I'll wait for someone to sign off on the final commit before merging.

**Review by irfansharif — APPROVED** [review]
Only change since last are the three extra generated files, right?

**Comment by rickystewart** [comment]
Yep. 👍

**Comment by RaduBerinde** [comment]
> figuring out which files to clean is the "smart" bit, that's some complex AI/ML work right there :D

Would it be feasible to have a special `files-to-delete` configuration somewhere that is manually updated whenever one of these files goes away? And both `make` and `dev` remove them before building? It happens rarely but when it does it's very annoying to switch between branches.

**Comment by irfansharif** [comment]
Would #74243 help? Assuming it's backported to release-21.1+.

**Comment by rickystewart** [comment]
YOLO

bors r=irfansharif

**Pull request merged**
